packages/core/src/dispatcher.ts

```diff
@@ -134,37 +153,107 @@ export default class Dispatcher extends EventEmitter {
-      const logMessage = { ...message, data: parsedMessageData }
-      this._log({ peer: this._peerId, event: 'received', topic: this.topic, message: logMessage })
-
-      const { typ, id, cid } = parsedMessageData
-      if (this._documents[id]) {
-        switch (typ) {
-          case MsgType.UPDATE:
-            if (typeof cid !== 'string') break
-            // add cache of cids here so that we don't emit event
-            // multiple times if we get the message more than once.
-            this._documents[id].emit('update', new CID(cid))
-            break
-          case MsgType.REQUEST:
-            this._documents[id].emit('tipreq')
-            break
-        }
-      }
+    if (message.from === this._peerId) {
+      return
+    }
+
+    // TODO: This is not a great way to handle the message because we don't
+    // don't know its type/contents. Ideally we can make this method generic
+    // against specific interfaces and follow IPFS specs for
+    // types (e.g. message data should be a buffer)
+    let parsedMessageData
+    if (typeof message.data === 'string') {
+      parsedMessageData = JSON.parse(message.data)
+    } else {
+      parsedMessageData = JSON.parse(new TextDecoder('utf-8').decode(message.data))
+    }
+    // TODO: handle signature and key buffers in message data
+    const logMessage = { ...message, data: parsedMessageData }
+    this._log({ peer: this._peerId, event: 'received', topic: this.topic, message: logMessage })
+
+    const { typ } = parsedMessageData
+    switch (typ) {
+      case MsgType.UPDATE:
+        await this._handleUpdateMessage(parsedMessageData)
+        break
+      case MsgType.QUERY:
+        await this._handleQueryMessage(parsedMessageData)
+        break
+      case MsgType.RESPONSE:
+        await this._handleResponseMessage(parsedMessageData)
+        break
+      default:
+        throw new Error("Unsupported message type: " + typ)
+    }
+
+  }
+
+  /**
+   * Handles an incoming Update message from the pub/sub topic.
+   * @param message
+   * @private
+   */
+  async _handleUpdateMessage(message: any): Promise<void> {
+    // TODO Add validation the message adheres to the proper format.
+
+    const { doc, tip } = message
+    if (!this._documents[doc]) {
+      return
+    }
+
+    // TODO: add cache of cids here so that we don't emit event
+    // multiple times if we get the message more than once.
+    this._documents[doc].emit('update', new CID(tip))
+
+    // TODO: Handle 'anchorService' if present in message
+  }
+
+  /**
+   * Handles an incoming Query message from the pub/sub topic.
+   * @param message
+   * @private
+   */
+  async _handleQueryMessage(message: any): Promise<void> {
+    // TODO Add validation the message adheres to the proper format.
+
+    const { doc: docId, id } = message
+    if (!this._documents[docId]) {
+      return
+    }
+
+    // TODO: Should we validate that the 'id' field is the correct hash of the rest of the message?
```

Comment:
I'm wondering if it's worth it to validate that the 'id' property is what we expect it to be, when receiving a QUERY message over the pubsub topic.  I'm leaning towards no, at least for now, as we don't actually *rely* on the id format for protocol correctness, it's just an optimization to allow deduping query messages.  Curious your thoughts on this though @oed @simonovic86

Reply:
yeah, I agree we dont' need to.